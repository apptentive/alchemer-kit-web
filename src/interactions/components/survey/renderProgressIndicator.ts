/**
 * Renders the step indicator at the bottom of the survey.
 *
 * @param {number} questionCount - The number of questions to track progress for
 * @returns {HTMLUListElement} - The DOM Node
 */
const renderProgressIndicator = (questionCount: number) => {
  const progressIndicatorWrapper = document.createElement('div');
  progressIndicatorWrapper.classList.add('apptentive-progress__wrapper');

  // Render progress bar if the number of questions is greater than 10
  if (questionCount > 10) {
    const progressBarElement = document.createElement('progress');
    progressBarElement.classList.add('apptentive-progress-bar');
    progressBarElement.max = 100;
    progressBarElement.value = 0;

    progressIndicatorWrapper.append(progressBarElement);
    return progressIndicatorWrapper;
  }

  // Otherwise render the step indicator with the number of questions
  const progressStepElement = document.createElement('ul');
  progressStepElement.classList.add('apptentive-step-indicator');

  // Create a page indicator based on how many total segments are configured for this survey
  for (let i = 0; i < questionCount; i++) {
    const indicatorElement = document.createElement('li');
    indicatorElement.classList.add('apptentive-step-indicator__item');

    if (i === 0) {
      indicatorElement.classList.add('apptentive-step-indicator__item--current');
    }

    progressStepElement.append(indicatorElement);
  }

  progressIndicatorWrapper.append(progressStepElement);
  return progressIndicatorWrapper;
};

export default renderProgressIndicator;
