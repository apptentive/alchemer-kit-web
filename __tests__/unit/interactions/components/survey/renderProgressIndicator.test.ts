import renderProgressIndicator from '../../../../../src/interactions/components/survey/renderProgressIndicator';

describe('renderProgressIndicator', () => {
  test('properly renders step progress indicator', () => {
    const element = renderProgressIndicator(5);

    expect(element).toMatchSnapshot();
  });

  test('properly renders indeterminate progress indicator', () => {
    const element = renderProgressIndicator(12);

    expect(element).toMatchSnapshot();
  });

  test('properly handles a question count of 0', () => {
    const element = renderProgressIndicator(0);

    expect(element).toMatchSnapshot();
  });
});
