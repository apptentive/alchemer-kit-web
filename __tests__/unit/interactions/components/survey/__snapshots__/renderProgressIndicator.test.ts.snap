// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderProgressIndicator properly handles a question count of 0 1`] = `
<div
  class="apptentive-progress__wrapper"
>
  <ul
    class="apptentive-step-indicator"
  />
</div>
`;

exports[`renderProgressIndicator properly renders indeterminate progress indicator 1`] = `
<div
  class="apptentive-progress__wrapper"
>
  <progress
    class="apptentive-progress-bar"
    max="100"
    value="0"
  />
</div>
`;

exports[`renderProgressIndicator properly renders step progress indicator 1`] = `
<div
  class="apptentive-progress__wrapper"
>
  <ul
    class="apptentive-step-indicator"
  >
    <li
      class="apptentive-step-indicator__item apptentive-step-indicator__item--current"
    />
    <li
      class="apptentive-step-indicator__item"
    />
    <li
      class="apptentive-step-indicator__item"
    />
    <li
      class="apptentive-step-indicator__item"
    />
    <li
      class="apptentive-step-indicator__item"
    />
  </ul>
</div>
`;
